File: algorithms/greedy/solutions.py
"""Greedy Algorithms — Essential Problem Solutions."""

from typing import List


def can_jump(nums: List[int]) -> bool:
    """Jump Game I — furthest reachable index tracking. O(n) time."""
    max_reach = 0
    for i, jump in enumerate(nums):
        if i > max_reach:
            return False
        max_reach = max(max_reach, i + jump)
    return True


def jump(nums: List[int]) -> int:
    """Jump Game II — minimum jumps to reach end. O(n) time."""
    jumps = 0
    current_end = farthest = 0
    for i in range(len(nums) - 1):
        farthest = max(farthest, i + nums[i])
        if i == current_end:
            jumps += 1
            current_end = farthest
    return jumps


def can_complete_circuit(gas: List[int], cost: List[int]) -> int:
    """Gas Station — running deficit & reset sweep. O(n) time."""
    if sum(gas) < sum(cost):
        return -1
    tank = start = 0
    for i in range(len(gas)):
        tank += gas[i] - cost[i]
        if tank < 0:
            start = i + 1
            tank = 0
    return start


def least_interval(tasks: List[str], n: int) -> int:
    """Task Scheduler — frequency bottleneck math. O(n) time."""
    from collections import Counter

    count = Counter(tasks)
    max_freq = max(count.values())
    max_count = sum(1 for v in count.values() if v == max_freq)
    part_count = max_freq - 1
    part_length = n + 1
    empty_slots = part_count * part_length
    available = len(tasks) - max_freq * max_count
    idle = max(0, empty_slots - available)
    return len(tasks) + idle


def candy(ratings: List[int]) -> int:
    """Candy — two-way pass (left-to-right & right-to-left). O(n) time."""
    n = len(ratings)
    candies = [1] * n
    for i in range(1, n):
        if ratings[i] > ratings[i - 1]:
            candies[i] = candies[i - 1] + 1
    for i in range(n - 2, -1, -1):
        if ratings[i] > ratings[i + 1]:
            candies[i] = max(candies[i], candies[i + 1] + 1)
    return sum(candies)


def erase_overlap_intervals(intervals: List[List[int]]) -> int:
    """Non-overlapping Intervals — earliest deadline first. O(n log n)."""
    intervals.sort(key=lambda x: x[1])
    count = 0
    prev_end = float("-inf")
    for start, end in intervals:
        if start >= prev_end:
            prev_end = end
        else:
            count += 1
    return count

File: algorithms/greedy/Solutions.java
import java.util.*;

/** Greedy Algorithms — Essential Problem Solutions. */
public class Solutions {

    /** Jump Game I — furthest reachable index tracking. O(n) time. */
    public static boolean canJump(int[] nums) {
        int maxReach = 0;
        for (int i = 0; i < nums.length; i++) {
            if (i > maxReach) return false;
            maxReach = Math.max(maxReach, i + nums[i]);
        }
        return true;
    }

    /** Jump Game II — minimum jumps to reach end. O(n) time. */
    public static int jump(int[] nums) {
        int jumps = 0, currentEnd = 0, farthest = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            farthest = Math.max(farthest, i + nums[i]);
            if (i == currentEnd) {
                jumps++;
                currentEnd = farthest;
            }
        }
        return jumps;
    }

    /** Gas Station — running deficit & reset sweep. O(n) time. */
    public static int canCompleteCircuit(int[] gas, int[] cost) {
        int totalGas = 0, totalCost = 0, tank = 0, start = 0;
        for (int i = 0; i < gas.length; i++) {
            totalGas += gas[i];
            totalCost += cost[i];
            tank += gas[i] - cost[i];
            if (tank < 0) {
                start = i + 1;
                tank = 0;
            }
        }
        return totalGas >= totalCost ? start : -1;
    }

    /** Task Scheduler — frequency bottleneck math. O(n) time. */
    public static int leastInterval(char[] tasks, int n) {
        int[] count = new int[26];
        for (char task : tasks) count[task - 'A']++;
        Arrays.sort(count);
        int maxFreq = count[25];
        int maxCount = 0;
        for (int c : count) if (c == maxFreq) maxCount++;
        int partCount = maxFreq - 1;
        int partLength = n + 1;
        int emptySlots = partCount * partLength;
        int available = tasks.length - maxFreq * maxCount;
        int idle = Math.max(0, emptySlots - available);
        return tasks.length + idle;
    }

    /** Candy — two-way pass (left-to-right & right-to-left). O(n) time. */
    public static int candy(int[] ratings) {
        int n = ratings.length;
        int[] candies = new int[n];
        Arrays.fill(candies, 1);
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) candies[i] = candies[i - 1] + 1;
        }
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1])
                candies[i] = Math.max(candies[i], candies[i + 1] + 1);
        }
        return Arrays.stream(candies).sum();
    }

    /** Non-overlapping Intervals — earliest deadline first. O(n log n). */
    public static int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[1]));
        int count = 0;
        int prevEnd = Integer.MIN_VALUE;
        for (int[] interval : intervals) {
            if (interval[0] >= prevEnd) {
                prevEnd = interval[1];
            } else {
                count++;
            }
        }
        return count;
    }
}
